homeassistant/components/vacuum/xiaomi.py

```diff
@@ -0,0 +1,375 @@
+        if entity_ids:
+            target_vacuums = [vac for vac in hass.data[PLATFORM].values()
+                              if vac.entity_id in entity_ids]
+        else:
+            target_vacuums = hass.data[PLATFORM].values()
+
+        update_tasks = []
+        for vacuum in target_vacuums:
+            yield from getattr(vacuum, method['method'])(**params)
+
+        for vacuum in target_vacuums:
+            if vacuum.should_poll:
+                update_coro = vacuum.async_update_ha_state(True)
+                update_tasks.append(update_coro)
+
+        if update_tasks:
+            yield from asyncio.wait(update_tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    for vacuum_service in SERVICE_TO_METHOD:
+        schema = SERVICE_TO_METHOD[vacuum_service].get(
+            'schema', VACUUM_SERVICE_SCHEMA)
+        hass.services.async_register(
+            DOMAIN, vacuum_service, async_service_handler,
+            description=descriptions.get(vacuum_service), schema=schema)
+
+
+class MiroboVacuum(VacuumDevice):
+    """Representation of a Xiaomi Vacuum cleaner robot."""
+
+    def __init__(self, hass, name, host, token):
+        """Initialize the Xiaomi vacuum cleaner robot handler."""
+        self.hass = hass
+        self._name = name
+        self._icon = ICON
+        self._host = host
+        self._token = token
+        self._vacuum = None
+
+        self.vacuum_state = None
+        self._is_on = False
+        self._available = False
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use for device."""
+        return self._icon
+
+    @property
+    def state(self) -> str:
+        """Return the state of the vacuum cleaner as a binary state."""
+        return STATE_ON if self.is_on else STATE_OFF
+
+    @property
+    def status(self):
+        """Return the status of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            return self.vacuum_state.state
+        return None
+
+    @property
+    def battery_level(self):
+        """Return the battery level of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            return self.vacuum_state.battery
+        return None
+
+    @property
+    def fanspeed(self):
+        """Return the fan speed of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            speed = self.vacuum_state.fanspeed
+            if speed in FANSPEED_VALUES:
+                return FANSPEED_LABELS[FANSPEED_VALUES.index(speed)]
+            return speed
+        return None
+
+    @property
+    def fanspeed_list(self):
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        return list(FANSPEED_LABELS)
```

Comment:
Even when not using dict/enum, the labels are already a list so no need to reconstruct it here.